### ISSUE TITLE
vs-project-load does not find tests in csproj that use PackageReference instead of package.config

### ISSUE THREAD

**Issue by ManfredLange** [nunit/vs-project-loader#16]
Problem: C# project file (*.csproj) may use <PackageReference> instead of <Reference> plus package.config to reference nuget package NUnit. If it does then vs-project-loader does not recognize the project of referencing NUnit and therefore does not include it and tests contained in it when using the solution file as a parameter.

The code in question is [here](https://github.com/nunit/vs-project-loader/blob/94e0418de688ef59ece06e4729f0ae48ed1cda7a/src/extension/VSSolution.cs#L192): It selects nodes named "Reference" but does not consider nodes named "PackageReference".

I believe there to be an easy fix. I might be able to create a pull request in the next few days with such a fix including one or more tests.

Steps to reproduce:
1. Using Visual Studio 2017 version 15.2 or 15.3 select "Add New Project"
2. Select "Installed" -> "Visual C#"
3. Select ".NET Framework 4.6.1" from the dropdown
4. Select template "Class Library (.NET Framework)" and click "OK"
5. For the new project select "Manage NuGet Packages..."
6. Find the "NUnit" package and install latest stable
7. Result: Project will not have a packages.config file. NUnit.framework will not be referenced as <Reference> in project file. Instead nuget package "NUnit" will be referenced as PackageReference in project file.

**Comment by CharliePoole** [comment]
Wow! I had no idea that code was in there. FYI, I simply uploaded the project mechanically, taking it from it's original location in the engine, before we made this a separate repo. I'll have to try to find the old code.

Here's the problem. The NUnit engine is supposed to be able to run all sorts of frameworks, provided somebody writes a driver for the new framework. So the project loader, in theory, ought to be returning any output assemblies it finds, leaving it up to the engine to decide - using the available drivers - whether it's an NUnit assembly. Still, we should research how and when this code got there. AFAIR it was not in NUnit V2, but I could be wrong.

An experiment you might try... what if you disable this check entirely? How does the console runner react when the solution includes non-test files.

**Comment by ManfredLange** [comment]
@CharliePoole I'll try the experiment you suggested. Thank you for the suggestion!

On one hand I can understand a line of reasoning that the project loader understands the file format and perhaps the engine shouldn't be bother with the specifics of a given format. Perhaps method "CheckProjectReferencesNunit()" should be in class VSProject, e.g. as "Project.ReferencesNunit() or something like that.

If the intention is to support different unit testing frameworks, then it could also be named "Project.ReferencesTestingFramework()" or "Project.MayContainTests()" if a test framework reference is detected. If a solution has 50 or 100 projects with half of them not containing tests, then perhaps it would be more efficient to load the project only once or discard it as soon as it can be determined with certainty that it doesn't contain tests.

Obviously the check, ie whether an assembly/project contains tests, can be made by looking at the assembly using a driver/adapter/etc. The vs-project-loader already loads the project files anyway and perhaps not having to even look at the assembly might have a performance benefit.

**Comment by ManfredLange** [comment]
@CharliePoole I can confirm that with the suggested experiment ie commenting out [line 129 in file VSSolution.cs](https://github.com/nunit/vs-project-loader/blob/94e0418de688ef59ece06e4729f0ae48ed1cda7a/src/extension/VSSolution.cs#L129) the tests are picked up correctly as well.

So we have at least the following two options to resolve this issue:
1. Remove the check altogether, ie lines 120 and method CheckProjectReferencesNunit()
2. Add code to also check for PackageReference nodes in method CheckProjectReferencesNunit()

Which way do you want to go in the overall context of NUnit and it's different parts?

**Comment by CharliePoole** [comment]
Sometimes when I review a draft that I've written, I find that I can make the text better by just crossing out whole lines of text. I love that. Works for code as well. 😄 

I think this is the job of the driver service in the engine, which only works with assemblies (i.e. after any projects are expanded). Each driver has it's own rules for recognizing that it knows how to deal with a particular assembly. The way we have tried to resolve the non-test assembly issue in the adapter is to pass a setting that says "just ignore any non-test assemblies" We could do the same thing when we have used the VS solution, which may include non-test assemblies. In other cases, the user has told us to run tests from a particular assembly, so not finding them is an error.

Bottom line, I'd take option 1 and just fix any problems that arise as a result.

**Comment by CharliePoole** [comment]
I'm still curious to know when and how we added that code though. I don't think it came over from V2.

**Comment by ManfredLange** [comment]
Shouldn't be to hard to find out how that code made it into this repo.

Regarding the fix: As you suggested I'll go with option 1 then when preparing the pull request.

**Comment by ManfredLange** [comment]
@CharliePoole There is a problem with just commenting out the check. Test FromSolutionWithNonNunitTestProject() fails in that case because without the check it will also add the assembly that does not reference NUnit. Therefore there are two subpackages each for the "Release" and "Debug" package.

Since you explained that vs-project-loader should not decide whether or not an assembly is worth loading - that is left to the engine or the drivers the engine uses respectively - I believe we should adjust test FromSolutionWithNonNunitTestProject() accordingly to expect 2 subpackages each for the "Release" and the "Debug" test packages.

All other tests are passing with commenting out line 129 in VSSolution.cs.

For the pull request I have an additional test for a solution plus project with a package reference. And I'll adjust test FromSolutionWithNonNunitTestProject() so that it expects two subpackages for each of the test packages.

Please confirm if this is in line with your vision. Thank you!

**Comment by CharliePoole** [comment]
It suits me. The test is testing what we actually do, which is fine, but I think we can change that.

I found where this came in: https://github.com/nunit/nunit/pull/1393

As you see, it's an old pull on the nunit project, from before we split the projects apart. I was concerned about it but agreed that it should go in as a temporary fix. I believe that we have now improved the engine and console to a point where it can be safely removed but you should run some manual tests with the console using non-test files in order to see whether the output you get is sufficiently friendly.

As I mentioned earlier, the VS Test Explorer gives us this same sort of unfiltered lists of assemblies and we now deal correctly with it. Should also be possible from the comand-line. I think that any package returned from VSSolution.cs should have the setting that tells us to ignore non-test assemblies - you'll have to look that up in the code. We should continue to give an error if you use a csproj, since that only has one assembly output and you should not be entering it on the command-line if it's not a test assembly.

**Comment by ManfredLange** [comment]
I've submitted a [pull request](https://github.com/nunit/vs-project-loader/pull/17) with a fix for this.

In addition to the tests I've also run this against a solution that contains a large number of assemblies without NUnit tests. From that it appears as if the NUnit engine handles it as expected when using flag skipnontestassemblies.

**Comment by ManfredLange** [comment]
Since the pull request for this issue has been merged into master, I'll close this issue.

**Comment by CharliePoole** [comment]
Yes... we usually use one of the automatic closing formats in our PR description - like "Fixes #nn". I didn't notice it wasn't there.
